Add controllers

uncle-sly opened the pull request:
Sprint 14

u-sly committed:
Создание структуры проекта: пакеты, интерфейсы, классы.

u-sly committed:
Добавлен ErrorHandlingControllerAdvice для обработки исключений, и классы для возврата сообщений исключений ErrorResponse, ValidationErrorResponse, ValidationViolation

Реализована часть касательно User
Контроллер, сервис, репозиторий. Сервис и Репозиторий через Интерфейс. Маппинг из модели в dto и обратно реализован через mapstruct и interface UserMapper, для чего в pom.xml добавлена зависимость mapstruct и в плагин maven-compiler-plugin конфигурации mapstruct-processor, lombok, lombok-mapstruct-binding.

Созданы маркерные интерфейсы OnCreate, OnUpdate.

u-sly committed:
исправление checkstyle

u-sly committed:
Реализация функциональности Item
добавлен ExceptionHandler для MissingRequestHeaderException, если заголовок пустой.
ItemController - обработка ендпоинтов вещей.
ItemServiceImpl - реализация логики приложения по работе с вещами.
InMemoryItemRepositoryImpl - реализованы методы для работы с хранилищем вещей.

ItemDto - DTO для отдачи клиенту наружу.
ItemMapper - реализован интерфейс через mapstruct как и в User, реализация генерируется автоматически.

ItemNotFoundException, ItemUpdateException - добавлены исключения для Item.

u-sly committed:
перемещение пакета validation из user в корень.

u-sly committed:
добавление логирования в метод create в UserController

u-sly committed:
добавление логирования с помощью AOP
добавлена зависимость spring-boot-starter-aop
добавлен класс LoggingAspect
настроено логирование публичных методов контроллеров UserController и ItemController.
Тесты Postman прошли успешно.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,59 @@
+    private final ItemService itemService;
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
```
Статус 200 OK является дефолтным) Можно было бы не указывать `@ResponseStatus(HttpStatus.OK)` над методами

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,59 @@
-}
+
+    //заголовок в константу
+    private static final String USER_ID = "X-Sharer-User-Id";
```
Это значение лучше вынести в отдельный класс с константами, так как использоваться будет не только здесь)

uncle-sly replied:
спасибо за рекомендацию)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,39 @@
+
+    @Size(max = 256, message = "Описание не может быть длиннее 256 символов.")
+    @NotBlank(groups = OnCreate.class, message = "name не может состоять из пробелов")
+    @NotEmpty(groups = OnCreate.class, message = "name не может быть пустым")
```
Эта проверка входит в `@NotBlank`, аннотация не требуется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,39 @@
+    private Long id;
+
+    @NotBlank(groups = OnCreate.class, message = "name не может состоять из пробелов")
+    @NotEmpty(groups = OnCreate.class, message = "name не может быть пустым")
```
Эта проверка входит в `@NotBlank`, аннотация не требуется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,39 @@
+    @NotEmpty(groups = OnCreate.class, message = "name не может быть пустым")
+    private String description;
+
+    @BooleanFlag
```
А для чего нам эта аннотация?)

uncle-sly replied:
ошибка, она нам не нужна, т.к. не участвует в валидации) 
NotNull будет достаточно)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,21 @@
+@Mapper
+public interface ItemMapper {
+
+    ItemMapper INSTANCE = Mappers.getMapper(ItemMapper.class);
```
Можно было бы представить этот класс в виде бина, так с ним работать будет удобнее)
`@Mapper(componentModel = "spring")`
https://www.baeldung.com/mapstruct

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/exception/ItemNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.practicum.shareit.item.exception;
+
+public class ItemNotFoundException extends RuntimeException {
```
Можно не делать отдельные исключения notFound. В больших проектах сущностей может быть сотни, не делать же под каждую отдельное исключение. Можно объявить одно общее `EntityNotFoundException` и передавать для информативности в его конструктор класс объекта сущности

uncle-sly replied:
супер, спасибо за идею)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,32 @@
+import ru.practicum.shareit.user.model.User;
+
 /**
  * TODO Sprint add-controllers.
```
Если функционал класса реализован - комментарий следует исключить. Либо же следует описать то, что остается реализовать. Аналогично в других местах

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,32 @@
+    private User owner;
+    private ItemRequest request;
+
+//    public Item(Long id, String name, String description, Boolean isAvailable) {
```
Если код не используется - следует его исключить

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,60 @@
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public UserDto create(@Validated (OnCreate.class) @RequestBody UserDto userDto) {
+        //        log.info("POST / Users --> Create User: {} - started", userDto);
```
Неиспользуемый код также следует исключить

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,60 @@
+    private final UserService userService;
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
```
Аннотация также не требуется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,32 @@
+    @NotNull(groups = OnUpdate.class)
+    private Long id;
+
+    @NotEmpty(groups = OnCreate.class)
```
`@NotEmpty` также не требуется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,32 @@
+    @NotBlank(groups = OnCreate.class)
+    private String name;
+
+    @NotEmpty(groups = OnCreate.class)
```
`@NotEmpty` также не требуется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,32 @@
+
+    @NotEmpty(groups = OnCreate.class)
+    @NotBlank(groups = OnCreate.class)
+    @Email(regexp = "^[a-zA-Z0-9_!#$%&’*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$", message = "некорретный email адрес", groups = OnCreate.class)
```
Хорошо было бы также указать группу обновления {...., ...}
Чтобы формат почты проверялся и при обновлении, если строка будет передана

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/exception/UserNotFoundException.java:
Также отдельный класс-исключение можно не вводить

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,19 @@
+import lombok.NoArgsConstructor;
+
+/**
+ * TODO Sprint add-controllers.
```
Комментарий также не требуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+        if (userRepository.checkEmail(userDto.getEmail()))
+            throw new UserEmailExistedException("Такой email уже используется.");
+
+        user.setName(userDto.getName());
```
А проверить строку на `null` и с помощью `isBlank`, чтобы убедиться, что она была передана?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+            throw new UserEmailExistedException("Такой email уже используется.");
+
+        user.setName(userDto.getName());
+        user.setEmail(userDto.getEmail());
```
А проверить строку на `null` и с помощью `isBlank`, чтобы убедиться, что она была передана?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public void delete(Long id) {
+        users.remove(id);
```
Не забываем удалить почту

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+        User user = userRepository.getById(id)
+                .orElseThrow(() -> new UserNotFoundException("Пользователь c ID - " + id + ", не найден."));
+
+        if (userRepository.checkEmail(userDto.getEmail()))
```
А если почта пользователя не изменилась?) Не возникнет ли ложная ошибка?)

uncle-sly replied:
действительно, пропустил этот момент.. поправил, спасибо

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,55 @@
+            emails.remove(currentUser.getEmail());
+            emails.add(user.getEmail());
+        }
+        currentUser.setName(user.getName());
```
Мы уже засетили значения на уровне сервиса) Действительно ли это требуется?)

uncle-sly replied:
нет, не нужно)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,55 @@
+
+    public User update(User user) {
+        User currentUser = users.get(user.getId());
+        if (checkEmail(currentUser.getEmail())) {
```
Валидация также была произведена на уровне сервиса

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+            throw new ItemUpdateException("Не владелец! Редактировать вешь, может только ее владелец.");
+        }
+
+        Item updatedItem = itemMapper.toItem(itemDto);
```
Было бы легче осуществлять обновление существующей вещи, не перенося значения полей и не пересоздавая объект)
Логика была бы примерно такой:
```
final String name = itemDto.getName();
if (name != null && !name.isBlank()) {
    item.setName(name);
}
....
```

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+        updatedItem.setAvailable(updatedItem.getAvailable() != null ? updatedItem.getAvailable() : item.getAvailable());
+        updatedItem.setOwner(item.getOwner());
+
+        return itemMapper.toItemDto(itemRepository.update(updatedItem));
```
Вызов метода `update`, в таком случае, не потребуется, так как мы изменяем существующий объект, ссылка на который уже лежит в мапе)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepositoryImpl.java:
```diff
@@ -0,0 +1,48 @@
+        return item;
+    }
+
+    public Item update(Item item) {
```
Этот метдо более не потребуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+        String searchText = String.format(text).trim().toLowerCase();
+
+        return itemMapper.toItemDtoList(!(text.isEmpty() || text.isBlank()) ? itemRepository.getSearchedItems(searchText) : List.of());
```
Проверки с помощью методов `isEmpty` и `isBlank` дублируют друг друга) Хорошо было бы выбрать более подходящий метод из них)

uncle-sly replied:
isBlank более подходит т.к. покрывает и isEmpty

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+        String searchText = String.format(text).trim().toLowerCase();
+
+        return itemMapper.toItemDtoList(!(text.isEmpty() || text.isBlank()) ? itemRepository.getSearchedItems(searchText) : List.of());
```
Почему бы проверку строки не разместить выше?) Если строка пуста, можно сразу же вернуть результат в виде пустого списка

uncle-sly replied:
логично, спасибо)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+        userRepository.getById(userId)
+                .orElseThrow(() -> new UserNotFoundException("Пользователь c ID - " + userId + ", не найден."));
+
+        String searchText = String.format(text).trim().toLowerCase();
```
А зачем здесь `String.format`?)

uncle-sly replied:
не нужен, лишний, т.к. форматирование я не использую тут)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepositoryImpl.java:
```diff
@@ -0,0 +1,48 @@
+
+    public List<Item> getSearchedItems(String text) {
+        return items.values().stream()
+                .filter(Item::getAvailable)
```
(предложение) Более эффективно было бы описать проверки в одном фильтре)
https://www.baeldung.com/java-streams-multiple-filters-vs-condition

uncle-sly replied:
класс, спасибо, учту)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить. Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

u-sly committed:
Исправление согласно комментариям в ревью:
* добавлен класс для констант Constants
* добавлен общее исключение для сущностей EntityNotFoundException с передачей в него класса сущности
* InMemoryItemRepositoryImpl, переделан фильтр метода getSearchedItems, удален метод update
* InMemoryUserRepositoryImp, метод update изменена логика и переименован в updateEmails
* Item удален код и коментарий
* ItemController, вынесена константа в отдельный класс, удалены статусы ОК.
* ItemDto, удалены лишние аннотации и коментарий
* ItemMapper, Mappers удален, добавлена componentModel = "spring" для использования как бин
* ItemRepository удален метод update
* ItemServiceImpl , добавлено использование EntityNotFoundException, изменена логика в update и getSearchedItems согласно замечаниям
* User удаление комментариев
* UserController удаление комментариев и статусов ОК
* UserDto удалены лишние аннотации, в email добавлена группа OnUpdate
* Mappers удален, добавлена componentModel = "spring" для использования как бин
* UserRepository метод update переименован в updateEmails
* UserServiceImpl добавлено использование EntityNotFoundException, изменена логика в update согласно замечаниям

Тесты Postman прошли успешно.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void delete(Long id) {
+        emails.remove(users.get(id).getEmail());
```
Можно записать в одну строку:
`emails.remove(users.remove(id).getEmail());`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+        User user = userRepository.getById(id)
+                .orElseThrow(() -> new EntityNotFoundException(User.class, " c ID = " + id + ", не найден."));
+
+        if (userRepository.checkEmail(userDto.getEmail()) && !user.getEmail().equals(userDto.getEmail()))
```
Можно было бы изменить порядок проверок, так как вторая более легкая, так как не нужно обращаться к хранилищу)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+
+    public void updateEmails(User user, UserDto userDto) {
+        emails.remove(user.getEmail());
```
Можно было бы проверить, что, если почты идентичны - удаление с дальнейшим обновлением - не требуется)

TyutterinYakov reviewed (approved):
Доброе утро! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного дня!🌅🎇

u-sly committed:
PR одобрен ревьювером.
Исправление согласно оставшимся комментариям от ревьювера:
* InMemoryItemRepositoryImpl
* UserServiceImpl
Тесты Postman прошли успешно.
